main.cpp

```diff
@@ -12,20 +12,34 @@
 #include <clang/Tooling/Tooling.h>
 #include <clang/Rewrite/Core/Rewriter.h>
 #include <clang/Frontend/CompilerInstance.h>
-
 using namespace clang;
 using namespace clang::ast_matchers;
 using namespace clang::tooling;
 
 class CastCallBack : public MatchFinder::MatchCallback {
 public:
-    CastCallBack(Rewriter& rewriter) {
-        // Your code goes here
+    CastCallBack(Rewriter& rewriter):r(rewriter) {
     };
 
     void run(const MatchFinder::MatchResult &Result) override {
-        // Your code goes here
+        auto E = Result.Nodes.getNodeAs<clang::CStyleCastExpr>("cast");
+        if((E==nullptr)||(E->getExprLoc().isMacroID()) ) return;
+        else {
+          auto b = E->getLParenLoc();
+          auto e = E->getRParenLoc();
+          auto variable_end = 
+          Lexer::getLocForEndOfToken(E->getSubExprAsWritten()->IgnoreImpCasts()->getEndLoc(), 0, *Result.SourceManager, LangOptions());

```

Comment:
Please describe why you are using here `getSubExprAsWritten()` and `IgnoreImpCasts()`.
What would be the difference with the use of `getSubExprAsWritten()->IgnoreImpCasts()`, `getSubExpr()->IgnoreImpCasts()` and `getSubExprAsWritten()`?

Reply:
If we add the following code
llvm::outs() << "getSubExpr(): " << E->getSubExpr()->getStmtClassName()<<"\n";
llvm::outs() << "getSubExprAsWritten(): " << E->getSubExprAsWritten()->getStmtClassName()<<"\n";
llvm::outs() << "getSubExpr()->ignoreImpCasts(): " << E->getSubExpr()->IgnoreImpCasts()->getStmtClassName()<<"\n";
llvm::outs() << "getSubExprAsWritten()->ignoreImpCasts(): " << E->getSubExprAsWritten()->IgnoreImpCasts()->getStmtClassName()<<"\n";
into the run function, as a result we can see the following output.
![image](https://user-images.githubusercontent.com/61185072/170833231-1af8c278-f1d2-43a7-a0f7-f48c17fc053b.png)
It turns out that using getSubExprAsWritten()->IgnoreImpCasts(), makes no sense here, and this part can be replaced either by getSubExprAsWritten() or getSubExpr()->ignoreImpCasts(). Using just getSubExpr(), will not lead to the expected result.
Moreover, according to the documentation on the Internet, we can explicitly specify the difference in these functions, and what they are used for:
getSubExpr() retrieves the first cast subexpression.
getSubExprAsWritten() retrieves the cast subexpression as it was written in the source
ignoreImpCasts() skips past any implicit casts which might surround this expression until reaching a fixed point. It skips the following AST nodes: ImplicitCastExpr, FullExpr.